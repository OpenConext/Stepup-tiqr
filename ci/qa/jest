#!/usr/bin/env bash

cd $(dirname $0)/../../

yarn jest
