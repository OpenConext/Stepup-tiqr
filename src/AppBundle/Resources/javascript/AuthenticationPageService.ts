import { StatusPollService } from './StatusPollService';
import { ComponentCollection } from './Component/ComponentCollection';
import { NotificationClient } from './Client/NotificationClient';
import { Component } from './Component/Component';

export class AuthenticationPageService {
  private allComponents: Component;

  constructor(private statusPollingService: StatusPollService,
              private notificationClient: NotificationClient,
              private spinnerComponent: Component,
              private qrComponent: Component,
              private otpFormComponent: Component,
              private challengeExpiredComponent: Component,
              private statusErrorComponent: Component,
              private notificationErrorComponent: Component) {
    this.allComponents = new ComponentCollection([
      this.spinnerComponent,
      this.qrComponent,
      this.otpFormComponent,
      this.challengeExpiredComponent,
      this.statusErrorComponent,
      this.notificationErrorComponent,
    ]);
  }

  /**
   * The user can scan the QR code, keep the polling alive.
   */
  public switchToManual(): void {
    this.allComponents.hide();
    this.qrComponent.show();
  }

  /**
   * Wait for the OTP te be filled.
   *
   * We don't stop polling, maybe they clicked wrong, and are still scanning the QR code.
   */
  public switchToOtp() {
    this.allComponents.hide();
    this.qrComponent.show();
    this.otpFormComponent.show();
  }

  /**
   * When status request to the API challengeExpired.
   *
   * Stop polling and ask user to retry.
   */
  public switchToChallengeHasExpired() {
    this.statusPollingService.stop();
    this.allComponents.hide();
    this.challengeExpiredComponent.show();
  }

  /**
   * Failed to send notification. Show message that it failed and give the user to the option to scan the QR code.
   */
  public switchToNotificationFailed() {
    this.allComponents.hide();
    this.qrComponent.show();
    this.notificationErrorComponent.show();
  }

  /**
   * There is no device registered to send a notification, show QR code instead.
   */
  public switchToNoDevice() {
    this.allComponents.hide();
    this.qrComponent.show();
  }

  /**
   * Do a request to send the push notification and start the polling service for the actual status.
   */
  public switchToPolling() {
    this.allComponents.hide();
    this.spinnerComponent.show();
    this.notificationClient.send(this.notificationReceivedHandler, this.notificationErrorHandler);
    this.statusPollingService.waitAndRequestStatus(this.statusReceivedHandler, this.statusErrorHandler);
  }

  /**
   * We got an status error.
   */
  public switchToStatusRequestError() {
    this.statusPollingService.stop();
    this.allComponents.hide();
    this.statusErrorComponent.show();
  }

  public reloadPage() {
    document.location.replace(document.location.href);
  }

  /**
   * This will handle the status given by the AuthenticationController.authenticationStatusAction.
   */
  private statusReceivedHandler = (status: string) => {
    switch (status) {
      case 'pending':
        this.statusPollingService.waitAndRequestStatus(this.statusReceivedHandler, this.statusErrorHandler);
        break;
      case 'challenge-expired':
        this.switchToChallengeHasExpired();
        break;
      case 'needs-refresh':
        this.reloadPage();
        break;
    }
  };

  /**
   * Handler when actual request to the API failed.
   */
  private statusErrorHandler = () => {
    this.switchToStatusRequestError();
  };

  /**
   * Status response handler for the notification.
   */
  private notificationReceivedHandler = (status: string) => {
    switch (status) {
      case 'success':
        // Do nothing, push notification is send successfully so we can keep on polling.
        break;
      case 'error':
        this.switchToNotificationFailed();
        break;
      case 'no-device':
        this.switchToNoDevice();
        break;
    }
  };

  /**
   * When the push notification failed. (Maybe the name resolvers timed out etc)
   */
  private notificationErrorHandler = () => {
    this.switchToNotificationFailed();
  };

}
