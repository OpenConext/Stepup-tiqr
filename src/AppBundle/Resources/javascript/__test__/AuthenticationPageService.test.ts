import 'jest';
import { AuthenticationPageService } from '../AuthenticationPageService';

describe('AuthenticationPageService', () => {

  function createTestContext() {
    const notificationClient = {
      send: jest.fn(),
    };
    const pollingService =  {
      waitAndRequestStatus: jest.fn(),
      stop: jest.fn(),
    };
    function createMockComponent() {
      return {
        show: jest.fn(),
        hide: jest.fn(),
      };
    }

    const spinnerComponent = createMockComponent();
    const qrComponent = createMockComponent();
    const otpFormComponent = createMockComponent();
    const challengeExpiredComponent = createMockComponent();
    const statusErrorComponent = createMockComponent();
    const notificationErrorComponent = createMockComponent();

    const authenticationPageService = new AuthenticationPageService(
      pollingService as any,
      notificationClient as any,
      spinnerComponent,
      qrComponent,
      otpFormComponent,
      challengeExpiredComponent,
      statusErrorComponent,
      notificationErrorComponent,
    );
    return {
      pollingService,
      authenticationPageService,
      notificationClient,
      spinnerComponent,
      qrComponent,
      otpFormComponent,
      challengeExpiredComponent,
      statusErrorComponent,
      notificationErrorComponent,
    };
  }

  let context = createTestContext();

  beforeEach(() => {
    context = createTestContext();
  });

  it('When switching to manual, the qr code should be shown', () => {
    context.authenticationPageService.switchToManual();
    expect(context.spinnerComponent.hide).toBeCalled();
    expect(context.spinnerComponent.show).not.toBeCalled();
    expect(context.qrComponent.show).toBeCalled();
  });

  describe('When notification failed', () => {
    beforeEach(() => {
      context = createTestContext();
      context.authenticationPageService.switchToOtp();
    });

    it('The qr code should be shown', () => {
      expect(context.spinnerComponent.hide).toBeCalled();
      expect(context.spinnerComponent.show).not.toBeCalled();
      expect(context.qrComponent.show).toBeCalled();
      expect(context.otpFormComponent.show).toBeCalled();
    });

    it('Polling should not be disabled', () => {
      expect(context.pollingService.stop).not.toBeCalled();
    });
  });

  describe('When challenge has expired', () => {
    beforeEach(() => {
      context = createTestContext();
      context.authenticationPageService.switchToChallengeHasExpired();
    });

    it('The warning should be shown', () => {
      expect(context.challengeExpiredComponent.show).toBeCalled();
    });

    it('Polling should be disabled', () => {
      expect(context.pollingService.stop).toBeCalled();
    });
  });

  describe('When notification failed', () => {

    beforeEach(() => {
      context = createTestContext();
      context.authenticationPageService.switchToNotificationFailed();
    });

    it('The warning and qr code should be shown', () => {
      expect(context.notificationErrorComponent.show).toBeCalled();
      expect(context.qrComponent.show).toBeCalled();
    });

    it('Polling should not be disabled', () => {
      expect(context.pollingService.stop).not.toBeCalled();
    });
  });
});
