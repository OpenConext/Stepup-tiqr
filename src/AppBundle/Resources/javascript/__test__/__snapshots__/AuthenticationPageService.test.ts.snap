// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticationPageService When push notification response is handled Should do nothing when push notification succeeded: notification succeeded 1`] = `
Object {
  "authenticationPageService": AuthenticationPageService {
    "allComponents": ComponentCollection {
      "components": Array [
        Object {
          "hide": [Function],
          "isVisible": [Function],
          "show": [Function],
        },
        Object {
          "hide": [Function],
          "isVisible": [Function],
          "show": [Function],
        },
        Object {
          "hide": [Function],
          "isVisible": [Function],
          "show": [Function],
        },
        Object {
          "hide": [Function],
          "isVisible": [Function],
          "show": [Function],
        },
        Object {
          "hide": [Function],
          "isVisible": [Function],
          "show": [Function],
        },
        Object {
          "hide": [Function],
          "isVisible": [Function],
          "show": [Function],
        },
      ],
    },
    "challengeExpiredComponent": Object {
      "hide": [Function],
      "isVisible": [Function],
      "show": [Function],
    },
    "notificationClient": Object {
      "send": [Function],
    },
    "notificationErrorComponent": Object {
      "hide": [Function],
      "isVisible": [Function],
      "show": [Function],
    },
    "notificationErrorHandler": [Function],
    "notificationReceivedHandler": [Function],
    "otpFormComponent": Object {
      "hide": [Function],
      "isVisible": [Function],
      "show": [Function],
    },
    "qrComponent": Object {
      "hide": [Function],
      "isVisible": [Function],
      "show": [Function],
    },
    "spinnerComponent": Object {
      "hide": [Function],
      "isVisible": [Function],
      "show": [Function],
    },
    "statusErrorComponent": Object {
      "hide": [Function],
      "isVisible": [Function],
      "show": [Function],
    },
    "statusErrorHandler": [Function],
    "statusPollingService": Object {
      "stop": [MockFunction],
      "waitAndRequestStatus": [MockFunction] {
        "calls": Array [
          Array [
            [Function],
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
    "statusReceivedHandler": [Function],
  },
  "challengeExpiredComponent": Object {
    "hide": [Function],
    "isVisible": [Function],
    "show": [Function],
  },
  "notificationClient": Object {
    "send": [Function],
  },
  "notificationErrorComponent": Object {
    "hide": [Function],
    "isVisible": [Function],
    "show": [Function],
  },
  "otpFormComponent": Object {
    "hide": [Function],
    "isVisible": [Function],
    "show": [Function],
  },
  "pollingService": Object {
    "stop": [MockFunction],
    "waitAndRequestStatus": [MockFunction] {
      "calls": Array [
        Array [
          [Function],
          [Function],
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    },
  },
  "qrComponent": Object {
    "hide": [Function],
    "isVisible": [Function],
    "show": [Function],
  },
  "spinnerComponent": Object {
    "hide": [Function],
    "isVisible": [Function],
    "show": [Function],
  },
  "statusErrorComponent": Object {
    "hide": [Function],
    "isVisible": [Function],
    "show": [Function],
  },
}
`;

exports[`AuthenticationPageService When push notification response is handled Should do nothing when push notification succeeded: notification succeeded 2`] = `
Object {
  "authenticationPageService": AuthenticationPageService {
    "allComponents": ComponentCollection {
      "components": Array [
        Object {
          "hide": [Function],
          "isVisible": [Function],
          "show": [Function],
        },
        Object {
          "hide": [Function],
          "isVisible": [Function],
          "show": [Function],
        },
        Object {
          "hide": [Function],
          "isVisible": [Function],
          "show": [Function],
        },
        Object {
          "hide": [Function],
          "isVisible": [Function],
          "show": [Function],
        },
        Object {
          "hide": [Function],
          "isVisible": [Function],
          "show": [Function],
        },
        Object {
          "hide": [Function],
          "isVisible": [Function],
          "show": [Function],
        },
      ],
    },
    "challengeExpiredComponent": Object {
      "hide": [Function],
      "isVisible": [Function],
      "show": [Function],
    },
    "notificationClient": Object {
      "send": [Function],
    },
    "notificationErrorComponent": Object {
      "hide": [Function],
      "isVisible": [Function],
      "show": [Function],
    },
    "notificationErrorHandler": [Function],
    "notificationReceivedHandler": [Function],
    "otpFormComponent": Object {
      "hide": [Function],
      "isVisible": [Function],
      "show": [Function],
    },
    "qrComponent": Object {
      "hide": [Function],
      "isVisible": [Function],
      "show": [Function],
    },
    "spinnerComponent": Object {
      "hide": [Function],
      "isVisible": [Function],
      "show": [Function],
    },
    "statusErrorComponent": Object {
      "hide": [Function],
      "isVisible": [Function],
      "show": [Function],
    },
    "statusErrorHandler": [Function],
    "statusPollingService": Object {
      "stop": [MockFunction],
      "waitAndRequestStatus": [MockFunction] {
        "calls": Array [
          Array [
            [Function],
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
    "statusReceivedHandler": [Function],
  },
  "challengeExpiredComponent": Object {
    "hide": [Function],
    "isVisible": [Function],
    "show": [Function],
  },
  "notificationClient": Object {
    "send": [Function],
  },
  "notificationErrorComponent": Object {
    "hide": [Function],
    "isVisible": [Function],
    "show": [Function],
  },
  "otpFormComponent": Object {
    "hide": [Function],
    "isVisible": [Function],
    "show": [Function],
  },
  "pollingService": Object {
    "stop": [MockFunction],
    "waitAndRequestStatus": [MockFunction] {
      "calls": Array [
        Array [
          [Function],
          [Function],
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    },
  },
  "qrComponent": Object {
    "hide": [Function],
    "isVisible": [Function],
    "show": [Function],
  },
  "spinnerComponent": Object {
    "hide": [Function],
    "isVisible": [Function],
    "show": [Function],
  },
  "statusErrorComponent": Object {
    "hide": [Function],
    "isVisible": [Function],
    "show": [Function],
  },
}
`;
